src/main/java/org/prebid/server/handler/NotificationEventHandler.java

```diff
@@ -0,0 +1,148 @@
+import org.prebid.server.analytics.model.HttpContext;
+import org.prebid.server.analytics.model.NotificationEvent;
+import org.prebid.server.exception.InvalidRequestException;
+import org.prebid.server.proto.request.EventNotificationRequest;
+import org.prebid.server.util.ResourceUtil;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Accepts notifications from browsers and mobile application for further processing by {@link AnalyticsReporter}
+ * and responding with tracking pixel when requested.
+ */
+public class NotificationEventHandler implements Handler<RoutingContext> {
+
+    private static final String TRACKING_PIXEL_PNG = "static/tracking-pixel.png";
+    private static final String TRACKING_PIXEL_JPG = "static/tracking-pixel.jpg";
+    private static final String VIEW_TYPE = "view";
+    private static final String WIN_TYPE = "win";
+    private static final String FORMAT_PARAMETER = "format";
+    private static final String JPG = "jpg";
+    private static final String PNG = "png";
+    private static final String JPG_HEADER = "image/jpeg";
+    private static final String PNG_HEADER = "image/png";
+
+    private final AnalyticsReporter analyticsReporter;
+    private final byte[] trackingPixelPng;
+    private final byte[] trackingPixelJpg;
+
+    private NotificationEventHandler(AnalyticsReporter analyticsReporter, byte[] trackingPixelPng,
+                                     byte[] trackingPixelJpg) {
+        this.analyticsReporter = analyticsReporter;
+        this.trackingPixelJpg = trackingPixelJpg;
+        this.trackingPixelPng = trackingPixelPng;
+    }
+
+    public static NotificationEventHandler create(AnalyticsReporter analyticsReporter) {
+        return new NotificationEventHandler(Objects.requireNonNull(analyticsReporter),
+                readTrackingPixel(TRACKING_PIXEL_PNG), readTrackingPixel(TRACKING_PIXEL_JPG));
+    }
+
+    private static byte[] readTrackingPixel(String path) {
+        try {
+            return ResourceUtil.readByteArrayFromClassPath(path);
+        } catch (IOException e) {
+            throw new IllegalArgumentException(String.format("Failed to load pixel image at %s", path), e);
+        }
+    }
+
+    @Override
+    public void handle(RoutingContext context) {
+        final EventNotificationRequest eventNotificationRequest;
+        try {
+            eventNotificationRequest = makeEventRequest(context.getBody());
+        } catch (InvalidRequestException ex) {
+            respondWithBadStatus(context, ex.getMessage());
+            return;
+        }
+
+        final NotificationEvent notificationEvent = new NotificationEvent(eventNotificationRequest.getType(),
+                eventNotificationRequest.getBidId(), eventNotificationRequest.getBidder());
+
+        analyticsReporter.processEvent(notificationEvent);
+
+        final Map<String, String> queryParams = HttpContext.from(context).getQueryParams();
```

Comment:
There is not need to create `HttpContext` for the sake of just reading a query parameter - it could be read from `context` directly.

Reply:
Has this been addressed somehow?